File: backend/src/auth/models/project.py
"""
Project models for storing user projects and files.
"""

from sqlmodel import SQLModel, Field, Relationship
from typing import Optional, List, TYPE_CHECKING
from uuid import UUID, uuid4
from datetime import datetime
from enum import Enum

if TYPE_CHECKING:
    from .sub_project import SubProject


class ProjectStatus(str, Enum):
    """Project status enumeration"""
    DRAFT = "draft"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class FileType(str, Enum):
    """File type enumeration"""
    LATEX = "latex"
    TIKZ = "tikz"
    MARKDOWN = "markdown"
    TEXT = "text"
    IMAGE = "image"
    PDF = "pdf"


class Project(SQLModel, table=True):
    """Project model for storing user projects (Mother Project)"""
    __tablename__ = "projects"
    
    id: UUID = Field(default_factory=uuid4, primary_key=True)
    user_id: UUID = Field(foreign_key="users.id", index=True)
    title: str = Field(max_length=255)
    description: Optional[str] = Field(default=None, max_length=1000)
    status: ProjectStatus = Field(default=ProjectStatus.DRAFT)
    
    # LaTeX specific fields (for combined output)
    latex_content: Optional[str] = Field(default=None)  # Main LaTeX content
    compiled_pdf_url: Optional[str] = Field(default=None)
    preview_image_url: Optional[str] = Field(default=None)
    
    # Metadata
    is_template: bool = Field(default=False)
    tags: Optional[str] = Field(default=None)  # Comma-separated tags
    
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Relationships
    files: List["ProjectFile"] = Relationship(back_populates="project")
    sub_projects: List["SubProject"] = Relationship(back_populates="project")


class ProjectFile(SQLModel, table=True):
    """File model for storing project files"""
    __tablename__ = "project_files"
    
    id: UUID = Field(default_factory=uuid4, primary_key=True)
    project_id: UUID = Field(foreign_key="projects.id", index=True)
    
    filename: str = Field(max_length=255)
    file_type: FileType = Field(default=FileType.LATEX)
    content: Optional[str] = Field(default=None)  # For text-based files
    file_url: Optional[str] = Field(default=None)  # For binary files (images, PDFs)
    file_size: Optional[int] = Field(default=None)  # Size in bytes
    
    # For ordering files in project
    order_index: int = Field(default=0)
    
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Relationships
    project: Optional[Project] = Relationship(back_populates="files")

File: frontend/types/project.ts
// Project types

export type ProjectStatus = 'draft' | 'in_progress' | 'completed' | 'archived';
export type FileType = 'latex' | 'tikz' | 'markdown' | 'text' | 'image' | 'pdf';
export type SubProjectType = 'table' | 'diagram' | 'imageToLatex' | 'handwrittenFlowchart' | 'document';

export interface ProjectFile {
  id: string;
  project_id: string;
  filename: string;
  file_type: FileType;
  content: string | null;
  file_url: string | null;
  file_size: number | null;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface SubProject {
  id: string;
  project_id: string;
  user_id: string;
  title: string;
  description: string | null;
  sub_project_type: SubProjectType;
  latex_code: string | null;
  editor_data: string | null;  // JSON string with editor state
  source_file_id: string | null;
  preview_image_url: string | null;
  compiled_pdf_url: string | null;
  is_completed: boolean;
  created_at: string;
  updated_at: string;
}

export interface SubProjectListItem {
  id: string;
  project_id: string;
  title: string;
  description: string | null;
  sub_project_type: SubProjectType;
  is_completed: boolean;
  preview_image_url: string | null;
  created_at: string;
  updated_at: string;
}

export interface SubProjectCreate {
  title?: string;
  description?: string;
  sub_project_type: SubProjectType;
  latex_code?: string;
  editor_data?: string;
  source_file_id?: string;
}

export interface SubProjectUpdate {
  title?: string;
  description?: string;
  latex_code?: string;
  editor_data?: string;
  source_file_id?: string;
  preview_image_url?: string;
  is_completed?: boolean;
}

export interface SubProjectAutoSave {
  latex_code?: string;
  editor_data?: string;
}

export interface SubProjectFileLink {
  id: string;
  sub_project_id: string;
  project_file_id: string;
  usage_type: string;
  created_at: string;
}

export interface Project {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  status: ProjectStatus;
  latex_content: string | null;
  compiled_pdf_url: string | null;
  preview_image_url: string | null;
  is_template: boolean;
  tags: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProjectCollaboratorInfo {
  user_id: string;
  name: string;
  email: string;
  role: string;
}

export interface Project {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  status: ProjectStatus;
  latex_content: string | null;
  compiled_pdf_url: string | null;
  preview_image_url: string | null;
  is_template: boolean;
  tags: string | null;
  owner_name: string;
  collaborators: ProjectCollaboratorInfo[];
  created_at: string;
  updated_at: string;
  files: ProjectFile[];
  sub_projects?: SubProjectListItem[];
}

export interface ProjectListItem {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  status: ProjectStatus;
  preview_image_url: string | null;
  tags: string | null;
  role?: 'owner' | 'collaborator';
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  title: string;
  description?: string;
  status?: ProjectStatus;
  latex_content?: string;
  tags?: string;
}

export interface ProjectUpdate {
  title?: string;
  description?: string;
  status?: ProjectStatus;
  latex_content?: string;
  compiled_pdf_url?: string;
  preview_image_url?: string;
  tags?: string;
}

export interface FileCreate {
  filename: string;
  file_type?: FileType;
  content?: string;
  file_url?: string;
  order_index?: number;
}

export interface FileUpdate {
  filename?: string;
  file_type?: FileType;
  content?: string;
  file_url?: string;
  order_index?: number;
}

export interface AutoSaveResponse {
  success: boolean;
  saved_at: string;
  project_id: string;
}

export interface SubProjectAutoSaveResponse {
  success: boolean;
  message: string;
  updated_at: string;
}
